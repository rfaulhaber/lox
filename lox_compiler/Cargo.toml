[package]
name = "lox_compiler"
version = "0.1.0"
edition = "2024"

[dependencies]
lox_source = { path = "../lox_source" }
lox_vm = { path = "../lox_vm" }
thiserror = "1"

[dev-dependencies]
insta = { version = "1.42.1", features = ["yaml"] }
