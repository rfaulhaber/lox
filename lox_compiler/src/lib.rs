use lox_bytecode::{Chunk, Op};
use lox_source::{
    ast::{
        decl::Decl,
        expr::{BinaryOperator, Expr, Identifier, Literal, LogicalOperator, Number, UnaryOperator},
        program::Program,
        stmt::Stmt,
        visitor::Visitor,
    },
    parser::{ParseError, Parser},
};
use thiserror::Error;

#[derive(Debug, Clone, Error)]
pub enum CompilerError {
    #[error("Encountered parser error {0}")]
    ParserError(#[from] ParseError),
}

pub type CompilerResult = Result<Chunk, CompilerError>;

pub struct Compiler {
    ast: Program,
    chunk: Chunk,
}

impl<'c> Compiler {
    pub fn new(source: Program) -> Self {
        Self {
            ast: source,
            chunk: Chunk::new(),
        }
    }

    pub fn new_from_source(source: &'c str) -> Result<Self, CompilerError> {
        let program = Parser::from_source(source).parse()?;

        Ok(Compiler::new(program))
    }

    pub fn compile(&mut self) -> Result<(), CompilerError> {
        let _ = self.visit_program(self.ast.clone())?;

        Ok(())
    }

    pub fn bytecode(self) -> Chunk {
        self.chunk
    }

    fn write_int(&mut self, int: i64) {
        let idx = self.chunk.add_int(int);
        self.chunk.add_op(Op::Integer(idx));
    }

    fn write_float(&mut self, float: f64) {
        let idx = self.chunk.add_float(float);
        self.chunk.add_op(Op::Float(idx));
    }

    fn write_string(&mut self, string: String) {
        let idx = self.chunk.add_string(string);
        self.chunk.add_op(Op::String(idx))
    }
}

impl Visitor for Compiler {
    type Value = Result<(), CompilerError>;

    fn visit_expr(&mut self, expr: Expr) -> Self::Value {
        match expr {
            Expr::Literal(l) => self.visit_literal(l),
            Expr::Unary(op, expr) => self.visit_unary_expr(op, *expr),
            Expr::Call(callee, arguments) => self.visit_call_expr(*callee, arguments),
            Expr::Binary(left, op, right) => self.visit_binary_expr(*left, op, *right),
            Expr::Logical(left, op, right) => self.visit_logical_expr(*left, op, *right),
            Expr::Grouping(expr) => self.visit_expr(*expr),
            Expr::Get(_, _) => todo!(),
            Expr::Set(_, _, _) => todo!(),
            Expr::Var(id) => {
                let idx = self.chunk.add_string(id.name);
                self.chunk.add_op(Op::GetGlobal(idx));
                Ok(())
            }
            Expr::Assignment(id, expr) => self.visit_assignment_expr(id, *expr),
        }
    }

    fn visit_unary_expr(&mut self, op: UnaryOperator, expr: Expr) -> Self::Value {
        let _ = self.visit_expr(expr)?;

        let opcode = match op {
            UnaryOperator::Neg => Op::Negate,
            UnaryOperator::Not => Op::Not,
        };

        self.chunk.add_op(opcode);

        Ok(())
    }

    fn visit_binary_expr(&mut self, left: Expr, op: BinaryOperator, right: Expr) -> Self::Value {
        let _ = self.visit_expr(left)?;
        let _ = self.visit_expr(right)?;

        let opcode: &[Op] = match op {
            BinaryOperator::Eq => &[Op::Equal],
            BinaryOperator::Neq => &[Op::Equal, Op::Not],
            BinaryOperator::Lt => &[Op::Less],
            BinaryOperator::Lte => &[Op::Less, Op::Not],
            BinaryOperator::Gt => &[Op::Greater],
            BinaryOperator::Gte => &[Op::Greater, Op::Not],
            BinaryOperator::Add => &[Op::Add],
            BinaryOperator::Sub => &[Op::Subtract],
            BinaryOperator::Mul => &[Op::Multiply],
            BinaryOperator::Div => &[Op::Divide],
        };

        for op in opcode {
            self.chunk.add_op(op.clone());
        }

        Ok(())
    }

    fn visit_literal(&mut self, literal: Literal) -> Self::Value {
        match literal {
            Literal::Number(Number::Float(f)) => {
                self.write_float(f);
            }
            Literal::Number(Number::Int(i)) => {
                self.write_int(i);
            }
            Literal::String(s) => self.write_string(s),
            Literal::Bool(true) => self.chunk.add_op(Op::True),
            Literal::Bool(false) => self.chunk.add_op(Op::False),
            Literal::Nil => self.chunk.add_op(Op::Nil),
        };

        Ok(())
    }

    fn visit_grouping_expr(&mut self, expr: Expr) -> Self::Value {
        self.visit_expr(expr)
    }

    fn visit_assignment_expr(&mut self, id: Identifier, expr: Expr) -> Self::Value {
        let _ = self.visit_expr(expr)?;

        let idx = self.chunk.add_string(id.name);
        self.chunk.add_op(Op::SetGlobal(idx));

        Ok(())
    }

    fn visit_logical_expr(&mut self, left: Expr, op: LogicalOperator, right: Expr) -> Self::Value {
        todo!()
    }

    fn visit_call_expr(&mut self, callee: Expr, arguments: Vec<Expr>) -> Self::Value {
        todo!()
    }

    fn visit_program(&mut self, program: Program) -> Self::Value {
        let result = program
            .declarations
            .iter()
            .map(|d| self.visit_declaration(d.clone()))
            .last();

        result.unwrap_or(Ok(()))
    }

    fn visit_declaration(&mut self, decl: Decl) -> Self::Value {
        match decl {
            Decl::Class(id, superclass, funcs) => {
                self.visit_class_delcaration(id, superclass, funcs)
            }
            Decl::Func(name, parameters, body) => {
                self.visit_func_declaration(name, parameters, body)
            }
            Decl::Var(id, expr) => {
                let _ = match expr {
                    Some(expr) => self.visit_expr(expr)?,
                    None => {
                        self.chunk.add_op(Op::Nil);
                    }
                };

                self.chunk.add_string(id.name);

                Ok(())
            }
            Decl::Stmt(stmt) => self.visit_stmt(stmt),
        }
    }

    fn visit_class_delcaration(
        &mut self,
        id: Identifier,
        superclass: Option<Identifier>,
        funcs: Vec<Decl>,
    ) -> Self::Value {
        todo!()
    }

    fn visit_stmt(&mut self, stmt: Stmt) -> Self::Value {
        match stmt {
            Stmt::Block(_) => todo!(),
            Stmt::Expr(expr) => {
                let _ = self.visit_expr(expr)?;
                self.chunk.add_op(Op::Pop);

                Ok(())
            }
            Stmt::Print(expr) => {
                let _ = self.visit_expr(expr)?;
                self.chunk.add_op(Op::Print);

                Ok(())
            }
            Stmt::Return(_) => todo!(),
            Stmt::If(_, _, _) => todo!(),
            Stmt::While(_, _) => todo!(),
        }
    }

    fn visit_block(&mut self, block: Vec<Decl>) -> Self::Value {
        todo!()
    }

    fn visit_if_stmt(&mut self, cond: Expr, stmt: Stmt, else_stmt: Option<Stmt>) -> Self::Value {
        todo!()
    }

    fn visit_while_stmt(&mut self, cond: Expr, body: Stmt) -> Self::Value {
        todo!()
    }

    fn visit_func_declaration(
        &mut self,
        name: Identifier,
        parameters: Vec<Identifier>,
        body: Stmt,
    ) -> Self::Value {
        todo!()
    }

    fn visit_return_stmt(&mut self, expr: Option<Expr>) -> Self::Value {
        todo!()
    }
}

#[cfg(test)]
mod test {
    use lox_bytecode::Op;

    use super::*;

    #[test]
    fn number_literals() {
        let input = "123;";
        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn grouping() {
        let input = "(123);";
        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn unary_negation() {
        let input = "-123;";
        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_op(Op::Negate);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn basic_add() {
        let input = "123 + 456;";

        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_int(456);
        expected.add_op(Op::Integer(1));
        expected.add_op(Op::Add);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn basic_sub() {
        let input = "123 - 456;";

        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_int(456);
        expected.add_op(Op::Integer(1));
        expected.add_op(Op::Subtract);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn basic_mul() {
        let input = "123 * 456;";

        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_int(456);
        expected.add_op(Op::Integer(1));
        expected.add_op(Op::Multiply);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn basic_div() {
        let input = "123 / 456;";

        let mut expected = Chunk::new();
        expected.add_int(123);
        expected.add_op(Op::Integer(0));
        expected.add_int(456);
        expected.add_op(Op::Integer(1));
        expected.add_op(Op::Divide);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn booleans() {
        let input = "!true;";

        let mut expected = Chunk::new();
        expected.add_op(Op::True);
        expected.add_op(Op::Not);
        expected.add_op(Op::Pop);

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode();

        assert_eq!(result.disassemble(), expected.disassemble());
    }

    #[test]
    fn snapshots() {
        let input = r#"var beverage = "cafe au lait";
var breakfast = "beignets with " + beverage;
print breakfast;"#;

        let mut compiler = Compiler::new_from_source(input).unwrap();
        let _ = compiler.compile().unwrap();
        let result = compiler.bytecode().disassemble();

        insta::assert_yaml_snapshot!(result);
    }
}
