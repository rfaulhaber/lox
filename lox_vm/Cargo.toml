[package]
name = "lox_vm"
version = "0.1.0"
edition = "2021"

[dependencies]
thiserror = "1"
lox_source = { path = "../lox_source" }
lox_bytecode = { path = "../lox_bytecode" }
lox_value = { path = "../lox_value" }
jiff = "0.2.4"
