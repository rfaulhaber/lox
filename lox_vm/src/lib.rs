use std::{collections::HashMap, io::Write};

use lox_bytecode;
use lox_value::{Value, ValueOperatorError};

use lox_bytecode::{Chunk, Op};
use thiserror::Error;

pub type InterpretResult = Result<(), InterpreterError>;

#[derive(Debug, Error)]
pub enum InterpreterError {
    #[error("Invalid operand provided for {0}: expected {1}")]
    InvalidOperand(String, String),
    #[error("Insufficient stack length for operation {0}: {1}")]
    InsufficientStackLengthForOperation(BinaryOp, usize),
    #[error("Arithmetic error")]
    ArithmeticError(#[from] ValueOperatorError),
    #[error("Cannot negate something that isn't a number ({0})")]
    NegateError(Value),
    #[error("Stack is empty")]
    EmptyStack,
    #[error("IO error")]
    IOError(#[from] std::io::Error),
    #[error("Missing value at index {0}")]
    NoValueAtIndex(usize),
    #[error("Undefined variable {0}")]
    UndefinedVariable(String),
    #[error("Local not found {0}")]
    LocalNotFound(usize),
}

#[derive(Debug)]
pub enum InterpreterMode {
    Normal,
    Debug,
}

#[derive(Debug, PartialEq)]
pub enum InterpreterState {
    Running,
    Finished,
}

#[derive(Debug, PartialEq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    Gt,
    Lt,
}

impl std::fmt::Display for BinaryOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                BinaryOp::Add => "+",
                BinaryOp::Sub => "-",
                BinaryOp::Mul => "*",
                BinaryOp::Div => "/",
                BinaryOp::Eq => "==",
                BinaryOp::Gt => ">",
                BinaryOp::Lt => "<",
            }
        )
    }
}

#[derive(Debug)]
pub struct Interpreter {
    chunk: Chunk,
    ip: usize,
    stack: Vec<Value>,
    globals: HashMap<String, Value>,
    mode: InterpreterMode,
}

impl Interpreter {
    pub fn new() -> Self {
        Interpreter::new_vm(InterpreterMode::Normal)
    }

    pub fn new_debug() -> Self {
        Interpreter::new_vm(InterpreterMode::Debug)
    }

    pub fn set_mode(&mut self, mode: InterpreterMode) {
        self.mode = mode;
    }

    pub fn eval(&mut self, chunk: Chunk) -> Result<InterpreterState, InterpreterError> {
        self.chunk = chunk;
        self.ip = 0;

        self.run()
    }

    fn new_vm(mode: InterpreterMode) -> Self {
        Interpreter {
            chunk: Chunk::new(),
            ip: 0,
            stack: Vec::new(),
            globals: HashMap::new(),
            mode,
        }
    }

    fn run(&mut self) -> Result<InterpreterState, InterpreterError> {
        loop {
            match self.step() {
                Ok(InterpreterState::Running) => continue,
                Ok(InterpreterState::Finished) => return Ok(InterpreterState::Finished),
                Err(e) => return Err(e),
            }
        }
    }

    fn step(&mut self) -> Result<InterpreterState, InterpreterError> {
        let op = self.next_op_and_advance();

        match op {
            Some(Op::Float(index)) => {
                let constant = self.chunk.float_at(index).unwrap();
                self.stack.push(Value::from(constant));
            }
            Some(Op::Integer(index)) => {
                let constant = self.chunk.int_at(index).unwrap();
                self.stack.push(Value::from(constant));
            }
            Some(Op::String(index)) => {
                let constant = self.chunk.string_at(index).unwrap();
                self.stack.push(Value::from(constant));
            }
            Some(Op::Return) | None => return Ok(InterpreterState::Finished),
            Some(Op::Negate) => match self.stack.pop() {
                Some(Value::Number(n)) => self.stack.push(Value::Number(-n)),
                Some(v) => return Err(InterpreterError::NegateError(v)),
                None => unreachable!("negate operation found without operand"),
            },
            Some(Op::Add) => {
                let res = self.binary_op(BinaryOp::Add)?;
                self.stack.push(res)
            }
            Some(Op::Subtract) => {
                let res = self.binary_op(BinaryOp::Sub)?;
                self.stack.push(res)
            }
            Some(Op::Multiply) => {
                let res = self.binary_op(BinaryOp::Mul)?;
                self.stack.push(res)
            }
            Some(Op::Divide) => {
                let res = self.binary_op(BinaryOp::Div)?;
                self.stack.push(res)
            }
            Some(Op::True) => self.stack.push(Value::Bool(true)),
            Some(Op::False) => self.stack.push(Value::Bool(false)),
            Some(Op::Nil) => self.stack.push(Value::Nil),
            Some(Op::Not) => match self.stack.pop() {
                Some(value) => self.stack.push(Value::Bool(value.is_falsy())),
                None => unreachable!("not operation found without operand"),
            },
            Some(Op::Equal) => {
                let binary_op = self.binary_op(BinaryOp::Eq)?;
                self.stack.push(binary_op)
            }
            Some(Op::Greater) => {
                let binary_op = self.binary_op(BinaryOp::Gt)?;
                self.stack.push(binary_op)
            }
            Some(Op::Less) => {
                let binary_op = self.binary_op(BinaryOp::Lt)?;
                self.stack.push(binary_op)
            }
            Some(Op::Print) => {
                let value = self.stack.pop();

                match value {
                    Some(v) => {
                        let mut stdout = std::io::stdout().lock();
                        writeln!(&mut stdout, "{}", v)?;
                    }
                    None => return Err(InterpreterError::EmptyStack),
                }
            }
            Some(Op::Pop) => {
                let _ = self.stack.pop();
            }
            Some(Op::DefineGlobal(index)) => {
                let value = self.stack.pop();
                let name = match self.chunk.string_at(index) {
                    Some(value) => value,
                    None => return Err(InterpreterError::NoValueAtIndex(index)),
                };

                match value {
                    Some(v) => {
                        self.globals.insert(name, v);
                    }
                    None => return Err(InterpreterError::EmptyStack),
                }
            }
            Some(Op::GetGlobal(index)) => {
                let name = match self.chunk.string_at(index) {
                    Some(value) => value,
                    None => return Err(InterpreterError::NoValueAtIndex(index)),
                };

                let value = match self.globals.get(&name) {
                    Some(value) => value,
                    None => return Err(InterpreterError::UndefinedVariable(name)),
                };

                self.stack.push(value.clone());
            }
            Some(Op::SetGlobal(index)) => {
                let name = match self.chunk.string_at(index) {
                    Some(value) => value,
                    None => return Err(InterpreterError::NoValueAtIndex(index)),
                };

                if !self.globals.contains_key(&name) {
                    return Err(InterpreterError::UndefinedVariable(name));
                }

                let value = match self.stack.pop() {
                    Some(value) => value,
                    None => return Err(InterpreterError::EmptyStack),
                };

                let _ = self.globals.insert(name, value);
            }
            Some(Op::GetLocal(index)) => {
                let value = self.stack.get(index);

                if value.is_none() {
                    return Err(InterpreterError::LocalNotFound(index));
                }

                self.stack.push(value.cloned().unwrap());
            }
            Some(Op::SetLocal(index)) => {
                let top = self.stack.get(index);

                if top.is_none() {
                    return Err(InterpreterError::EmptyStack);
                }

                self.stack[index] = top.cloned().unwrap();
            }
            Some(Op::JumpIfFalse(pos)) => {
                let value = self.stack.last();

                let is_falsy = match value {
                    Some(v) => v.is_falsy(),
                    None => return Err(InterpreterError::EmptyStack),
                };

                if is_falsy {
                    self.ip += pos;
                }
            }
            Some(Op::Jump(pos)) => {
                self.ip += pos;
            }
            Some(Op::Loop(pos)) => {
                self.ip -= pos;
            }
        }

        Ok(InterpreterState::Running)
    }

    fn next_op(&self) -> Option<Op> {
        self.chunk.code_at(self.ip).cloned()
    }

    fn next_op_and_advance(&mut self) -> Option<Op> {
        let op = self.next_op();

        if op.is_some() {
            self.ip = self.ip + 1;
        }

        op
    }

    fn binary_op(&mut self, op: BinaryOp) -> Result<Value, InterpreterError> {
        if self.stack.len() < 2 {
            return Err(InterpreterError::InsufficientStackLengthForOperation(
                op,
                self.stack.len(),
            ));
        }

        let b = self.stack.pop().unwrap();
        let a = self.stack.pop().unwrap();

        let value = match op {
            BinaryOp::Add => (a + b)?,
            BinaryOp::Sub => (a - b)?,
            BinaryOp::Mul => (a * b)?,
            BinaryOp::Div => (a / b)?,
            BinaryOp::Eq => Value::Bool(a == b),
            BinaryOp::Gt => Value::Bool(a > b),
            BinaryOp::Lt => Value::Bool(a < b),
        };

        Ok(value)
    }
}

impl Default for Interpreter {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn stack_calculation() {
        let mut code = Chunk::new();
        code.push_float(1.2);
        code.push_float(3.4);

        code.add_op(Op::Add);

        let mut vm = Interpreter::new();
        vm.chunk = code;

        let _ = vm.step();
        let _ = vm.step();
        let _ = vm.step();

        assert_eq!(vm.stack.first().unwrap(), &Value::from(4.6));
    }

    #[test]
    fn string_concat() {
        let mut code = Chunk::new();
        code.push_string("foo".into());
        code.push_string("bar".into());
        code.push_string("baz".into());

        code.add_op(Op::Add);
        code.add_op(Op::Add);

        let mut vm = Interpreter::new();
        vm.chunk = code;

        let _ = vm.run();

        assert_eq!(
            vm.stack.first().unwrap(),
            &Value::from(String::from("foobarbaz"))
        )
    }
}
