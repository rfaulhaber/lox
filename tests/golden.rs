use std::path::{Path, PathBuf};

use lox_compiler::Compiler;
use lox_vm::Interpreter;

use insta::glob;

fn run_test(file_path: &Path) -> String {
    use std::fs;

    let in_contents = fs::read_to_string(file_path).expect("Should have been able to read in file");

    let mut actual_out = Vec::<u8>::new();

    let bytecode = Compiler::new_from_source(&*in_contents)
        .expect("Could not create compiler")
        .compile()
        .expect("Could not compile code");

    let mut interpreter = Interpreter::new_with_writer(&mut actual_out);

    let _ = interpreter.eval(bytecode).expect("Could not evaluate file");

    String::from_utf8_lossy(&actual_out).into()
}

#[test]
fn golden() {
    // I know that /technically/ these aren't "golden" tests, however `insta` is
    // functionally exactly what I'd want for golden tests, so I just reuse it
    glob!("fixtures/*.lox", |path| {
        println!("running test for {:?}", path);
        insta::assert_yaml_snapshot!(run_test(path));
    });
}
