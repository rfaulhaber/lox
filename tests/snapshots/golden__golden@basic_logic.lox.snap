---
source: tests/golden.rs
assertion_line: 33
expression: run_test(path)
input_file: tests/fixtures/basic_logic.lox
---
"\"hi\"\n\"yes\"\n"
